paimon-core/src/main/java/org/apache/paimon/utils/SnapshotManager.java

```diff
@@ -118,8 +118,20 @@ public Path snapshotDirectory() {
     }
 
     public Snapshot snapshot(long snapshotId) {
-        Path snapshotPath = snapshotPath(snapshotId);
-        return Snapshot.fromPath(fileIO, snapshotPath);
+        try {
+            Path snapshotPath = snapshotPath(snapshotId);
+            return Snapshot.fromPath(fileIO, snapshotPath);
+        } catch (Exception e) {
+            String errorMessage =
+                    String.format(
+                            "Failed to read snapshot file for snapshot %s. "
+                                    + "It might have been expired by other jobs operating on this table. "
+                                    + "In this case, you can avoid concurrent modification issues by configuring "
+                                    + "write-only = true and use a dedicated compaction job, or configuring "
+                                    + "different expiration thresholds for different jobs.",
+                            snapshotId);
+            throw new IllegalStateException(errorMessage, e);
```

Comment:
This is a bad idea, we should not wrap exception here. FileNotException will be hided.

You can still modify expire class.

Reply:
The FileNotFoundException has already been wrapped with RuntimeException in Snapshot#fromPath. I'll modify this method to avoid wrapping FileNotFoundException twice.